src/battle_util.c

```diff
@@ -1641,83 +1642,154 @@ static void TryToRevertMimicryAndFlags(void)
+    ENDTURN_ORDER,
+    ENDTURN_VARIOUS,
+    ENDTURN_WEATHER,
+    ENDTURN_WEATHER_FORM, // Temporarily
+    ENDTURN_WEATHER_DAMAGE,
+    ENDTURN_EMERGENCY_EXIT_1,
+    ENDTURN_AFFECTION,
+    ENDTURN_FUTURE_SIGHT,
+    ENDTURN_DOOM_DESIRE,
+    ENDTURN_WISH,
+    ENDTURN_FIRST_EVENT_BLOCK,
+    ENDTURN_EMERGENCY_EXIT_2,
+    ENDTURN_AQUA_RING,
+    ENDTURN_INGRAIN,
+    ENDTURN_LEECH_SEED,
+    ENDTURN_POISON,
+    ENDTURN_BURN,
+    ENDTURN_FROSTBITE,
+    ENDTURN_NIGHTMARE,
+    ENDTURN_CURSE,
+    ENDTURN_WRAP,
+    ENDTURN_SALT_CURE,
+    ENDTURN_OCTOLOCK,
+    ENDTURN_SYRUP_BOMB,
+    ENDTURN_TAUNT,
+    ENDTURN_TORMENT,
+    ENDTURN_ENCORE,
+    ENDTURN_DISABLE,
+    ENDTURN_MAGNET_RISE,
+    ENDTURN_TELEKINESIS,
+    ENDTURN_HEAL_BLOCK,
+    ENDTURN_EMBARGO,
+    ENDTURN_YAWN,
+    ENDTURN_PERISH_SONG,
+    ENDTURN_ROOST,
+    ENDTURN_EMERGENCY_EXIT_3,
+    ENDTURN_SECOND_EVENT_BLOCK,
+    ENDTURN_TRICK_ROOM,
+    ENDTURN_GRAVITY,
+    ENDTURN_WATER_SPORT,
+    ENDTURN_MUD_SPORT,
+    ENDTURN_WONDER_ROOM,
+    ENDTURN_MAGIC_ROOM,
+    ENDTURN_TERRAIN,
+    ENDTURN_THIRD_EVENT_BLOCK,
+    ENDTURN_EMERGENCY_EXIT_4,
+    ENDTURN_ABILITIES,
+    ENDTURN_FOURTH_EVENT_BLOCK,
+    ENDTURN_COUNT,
+};
 
-    for (gBattlerAttacker = 0; gBattlerAttacker < gBattlersCount && gAbsentBattlerFlags & gBitTable[gBattlerAttacker]; gBattlerAttacker++)
-    {
-    }
-    for (gBattlerTarget = 0; gBattlerTarget < gBattlersCount && gAbsentBattlerFlags & gBitTable[gBattlerTarget]; gBattlerTarget++)
+enum FirstEventBlock
+{
+    FIRST_EVENT_BLOCK_GMAX_MOVE_RESIDUAL,
+    FIRST_EVENT_BLOCK_SEA_OF_FIRE_DAMAGE,
+    FIRST_EVENT_BLOCK_THRASH,
+    FIRST_EVENT_BLOCK_GRASSY_TERRAIN_HEAL,
+    FIRST_EVENT_BLOCK_ABILITIES,
+    FIRST_EVENT_BLOCK_HEAL_ITEMS,
+};
+
+enum SecondEventBlock
+{
+    SECOND_EVENT_BLOCK_REFLECT,
+    SECOND_EVENT_BLOCK_LIGHT_SCREEN,
+    SECOND_EVENT_BLOCK_SAFEGUARD,
+    SECOND_EVENT_BLOCK_MIST,
+    SECOND_EVENT_BLOCK_TAILWIND,
+    SECOND_EVENT_BLOCK_LUCKY_CHANT,
+    SECOND_EVENT_BLOCK_RAINBOW,
+    SECOND_EVENT_BLOCK_SEA_OF_FIRE,
+    SECOND_EVENT_BLOCK_SWAMP,
+    SECOND_EVENT_BLOCK_AURORA_VEIL,
+};
+
+enum ThirdEventBlock
+{
+    THIRD_EVENT_BLOCK_UPROAR,
+    THIRD_EVENT_BLOCK_ABILITIES, // Bad Dreams, Ball Fetch, Harvest, Moody/ Pickup / Slow Start / Speed Boost
+    THIRD_EVENT_BLOCK_ITEMS, // Flame Orb, Sticky Barb, Toxic Orb, White Herb
+};
+
+enum FourthEventBlock
+{
+    FOURTH_EVENT_BLOCK_HUNGER_SWITCH,
+    FOURTH_EVENT_BLOCK_EJECT_PACK,
+};
+
+#define MAGIC_GUARD_CHECK \
+if (ability == ABILITY_MAGIC_GUARD) \
+{\
+    RecordAbilityBattle(battler, ability);\
+    gBattleStruct->turnEffectsBattlerId++;\
+            break;\
+}
+
+static inline u32 TryEmergencyExit(u32 battler)
+{
+    u32 ability = GetBattlerAbility(battler);
+    if (ability == ABILITY_EMERGENCY_EXIT || ability == ABILITY_WIMP_OUT)
     {
+        if (AbilityBattleEffects(ABILITYEFFECT_ENDTURN, battler, 0, 0, 0))
+            return 1;
     }
+    return 0;
+}
```

Comment:
**suggestion:** Should this return a `bool32` instead?

Reply:
yep